{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import os\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "import matplotlib.pyplot as plt\n",
    "from scanpy.metrics.specificity.plot import marker_genes_distribution, one_v_max_genelist\n",
    "import numpy as np\n",
    "from scanpy.metrics.specificity.get_data import get_average_celltype_counts,get_spe\n",
    "from scanpy.metrics.specificity.compute import one_v_max_matrix\n",
    "from joblib import dump, load\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import copy\n",
    "import upsetplot\n",
    "os.chdir('C:\\\\Users\\\\ipmc\\\\Documents\\\\Metrics_results')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_PATH = r'data/'\n",
    "MODEL_PATH = r'models/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "barbry_file = 'HCA_Barbry_Grch38_Raw_filter_Norm.h5ad'\n",
    "meyer_file = 'lung.cellxgene.h5ad'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ipmc\\.conda\\envs\\scanpypip\\lib\\site-packages\\anndata\\compat\\__init__.py:158: FutureWarning: Moving element from .uns['neighbors']['distances'] to .obsp['distances'].\n",
      "\n",
      "This is where adjacency matrices should go now.\n",
      "  warn(\n",
      "C:\\Users\\ipmc\\.conda\\envs\\scanpypip\\lib\\site-packages\\anndata\\compat\\__init__.py:158: FutureWarning: Moving element from .uns['neighbors']['connectivities'] to .obsp['connectivities'].\n",
      "\n",
      "This is where adjacency matrices should go now.\n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "barbry = anndata.read_h5ad(DATA_PATH + 'HCA_Barbry_Grch38_Raw_filter_Norm.h5ad')\n",
    "meyer = anndata.read_h5ad(DATA_PATH + 'lung.cellxgene.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=np.array(barbry.X.todense()[:,0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(barbry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL669831.5</th>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>3004.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.061341</td>\n",
       "      <td>1.173543</td>\n",
       "      <td>-0.641477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>964.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.027584</td>\n",
       "      <td>1.389954</td>\n",
       "      <td>-0.191512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM41C</th>\n",
       "      <td>ENSG00000230368</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>6935.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.197905</td>\n",
       "      <td>1.660163</td>\n",
       "      <td>0.370307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>16862.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.339094</td>\n",
       "      <td>1.345867</td>\n",
       "      <td>-0.283180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHN1</th>\n",
       "      <td>ENSG00000187583</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>1767.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.033787</td>\n",
       "      <td>1.164980</td>\n",
       "      <td>-0.659280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>5882.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.128567</td>\n",
       "      <td>1.306934</td>\n",
       "      <td>-0.364129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>765.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.023429</td>\n",
       "      <td>1.424258</td>\n",
       "      <td>-0.120188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL354822.1</th>\n",
       "      <td>ENSG00000278384</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>795.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.017429</td>\n",
       "      <td>1.275135</td>\n",
       "      <td>-0.430245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004556.1</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>3645.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.101551</td>\n",
       "      <td>1.655506</td>\n",
       "      <td>0.360624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ucagenomix-cellranger-GRCh38-3.0.0</td>\n",
       "      <td>3233.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.068309</td>\n",
       "      <td>1.267142</td>\n",
       "      <td>-0.446864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14349 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids    feature_types  \\\n",
       "AL669831.5  ENSG00000237491  Gene Expression   \n",
       "LINC00115   ENSG00000225880  Gene Expression   \n",
       "FAM41C      ENSG00000230368  Gene Expression   \n",
       "NOC2L       ENSG00000188976  Gene Expression   \n",
       "PLEKHN1     ENSG00000187583  Gene Expression   \n",
       "...                     ...              ...   \n",
       "AC007325.4  ENSG00000278817  Gene Expression   \n",
       "AC007325.2  ENSG00000277196  Gene Expression   \n",
       "AL354822.1  ENSG00000278384  Gene Expression   \n",
       "AC004556.1  ENSG00000276345  Gene Expression   \n",
       "AC240274.1  ENSG00000271254  Gene Expression   \n",
       "\n",
       "                                        genome  n_counts  highly_variable  \\\n",
       "AL669831.5  ucagenomix-cellranger-GRCh38-3.0.0    3004.0            False   \n",
       "LINC00115   ucagenomix-cellranger-GRCh38-3.0.0     964.0            False   \n",
       "FAM41C      ucagenomix-cellranger-GRCh38-3.0.0    6935.0            False   \n",
       "NOC2L       ucagenomix-cellranger-GRCh38-3.0.0   16862.0            False   \n",
       "PLEKHN1     ucagenomix-cellranger-GRCh38-3.0.0    1767.0            False   \n",
       "...                                        ...       ...              ...   \n",
       "AC007325.4  ucagenomix-cellranger-GRCh38-3.0.0    5882.0            False   \n",
       "AC007325.2  ucagenomix-cellranger-GRCh38-3.0.0     765.0            False   \n",
       "AL354822.1  ucagenomix-cellranger-GRCh38-3.0.0     795.0            False   \n",
       "AC004556.1  ucagenomix-cellranger-GRCh38-3.0.0    3645.0            False   \n",
       "AC240274.1  ucagenomix-cellranger-GRCh38-3.0.0    3233.0            False   \n",
       "\n",
       "               means  dispersions  dispersions_norm  \n",
       "AL669831.5  0.061341     1.173543         -0.641477  \n",
       "LINC00115   0.027584     1.389954         -0.191512  \n",
       "FAM41C      0.197905     1.660163          0.370307  \n",
       "NOC2L       0.339094     1.345867         -0.283180  \n",
       "PLEKHN1     0.033787     1.164980         -0.659280  \n",
       "...              ...          ...               ...  \n",
       "AC007325.4  0.128567     1.306934         -0.364129  \n",
       "AC007325.2  0.023429     1.424258         -0.120188  \n",
       "AL354822.1  0.017429     1.275135         -0.430245  \n",
       "AC004556.1  0.101551     1.655506          0.360624  \n",
       "AC240274.1  0.068309     1.267142         -0.446864  \n",
       "\n",
       "[14349 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barbry.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[0.03265522, 0.01286653, 0.09192908, ..., 0.00902156, 0.04433792,\n",
       "         0.03552148]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barbry.X.todense().mean(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Détermination des gènes marqueurs selon plusieurs méthodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On va tester la plupart des méthodes de ranking_genes_group.\n",
    "Pour ce qui est de la spécificité, on peut en faire plusieurs cat. Dans un premier temps, on garde les high et les unique et on regarde ce que ça donne (combien yen a etc...). On pourra ensuite jeter un coup d'oeil aux multi spé mais yaura surement besoin de la métrique magique."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ipmc\\.conda\\envs\\scanpypip\\lib\\site-packages\\anndata\\_core\\anndata.py:1192: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if is_string_dtype(df[key]) and not is_categorical(df[key])\n"
     ]
    }
   ],
   "source": [
    "sc.tl.rank_genes_groups(barbry, 'CellType', method='t-test')\n",
    "#sc.pl.rank_genes_groups(barbry, n_genes=20, sharey=False)\n",
    "\n",
    "ttest_de =  copy.deepcopy(barbry.uns['rank_genes_groups'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(barbry, 'CellType', method='wilcoxon')\n",
    "#sc.pl.rank_genes_groups(barbry, n_genes=20, sharey=False)\n",
    "\n",
    "wilco_de = copy.deepcopy(barbry.uns['rank_genes_groups'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.tl.rank_genes_groups(barbry, 'CellType', method='logreg')\n",
    "#sc.pl.rank_genes_groups(barbry, n_genes=20, sharey=False)\n",
    "\n",
    "#logreg_de = copy.deepcopy(barbry.uns['rank_genes_groups'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype((numpy.record, [('AT1', 'O'), ('AT2', 'O'), ('B cells', 'O'), ('Basal', 'O'), ('Brush cells', 'O'), ('Cycling Basal', 'O'), ('Dendritic', 'O'), ('Deuterosomal', 'O'), ('Endothelial', 'O'), ('Fibroblast', 'O'), ('Ionocyte', 'O'), ('LT/NK', 'O'), ('Macrophage', 'O'), ('Mast cells', 'O'), ('Monocyte', 'O'), ('Multiciliated', 'O'), ('Multiciliated N', 'O'), ('PNEC', 'O'), ('Pericyte', 'O'), ('Plasma cells', 'O'), ('Precursor', 'O'), ('SMG Goblet', 'O'), ('Secretory', 'O'), ('Secretory N', 'O'), ('Serous', 'O'), ('Smooth muscle', 'O'), ('Suprabasal', 'O'), ('Suprabasal N', 'O')]))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barbry.uns['rank_genes_groups']['names'].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dict_to_df(DE_dico, n_genes):\n",
    "    celltypes = barbry.obs['CellType'].cat.categories\n",
    "    df = pd.DataFrame(index = np.linspace(1,n_genes,n_genes)  , columns = celltypes )\n",
    "    for celltype in celltypes:\n",
    "        df[celltype] = DE_dico['names'][celltype][0:n_genes]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "wilco30 = dict_to_df(wilco_de,30)\n",
    "ttest30 = dict_to_df(ttest_de,30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ipmc\\.conda\\envs\\scanpypip\\lib\\site-packages\\pandas\\core\\frame.py:1485: FutureWarning: Using short name for 'orient' is deprecated. Only the options: ('dict', list, 'series', 'split', 'records', 'index') will be used in a future version. Use one of the above to silence this warning.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "wilco30_dict = wilco30.to_dict(orient='lines')\n",
    "ttest30_dict = ttest30.to_dict(orient='lines')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load specificity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This should now be included to scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_max_celltype(gene,adata,partition_key=\"CellType\"):\n",
    "    get_average_celltype_counts(adata , partition_key = partition_key)\n",
    "    gene_index = int(np.where(adata.var.index == gene)[0])\n",
    "    max_celltype_index = np.argmax(adata.varm['ave_celltype_counts_' + partition_key][gene_index,:])\n",
    "    max_celltype = list(adata.obs[partition_key].cat.categories)[max_celltype_index]\n",
    "    return(max_celltype)\n",
    "\n",
    "def full_spe_matrix(adata, partition_key: str = 'CellType') :\n",
    "    shannon = get_spe(adata, spe_metric='shannon', partition_key=partition_key)\n",
    "    tau = get_spe(adata, spe_metric='tau', partition_key=partition_key)\n",
    "    gini = get_spe(adata, spe_metric='gini', partition_key=partition_key)\n",
    "    one_v_max = one_v_max_matrix(adata,partition_key=partition_key)\n",
    "    spe_matrix = pd.DataFrame(columns=['gene',\n",
    "                                       'most expressed celltype',\n",
    "                                       'one_v_max',\n",
    "                                       'shannon',\n",
    "                                       'tau',\n",
    "                                       'gini'])\n",
    "    for gene in adata.var.index :\n",
    "        spe_matrix.loc[gene,:] = [gene,\n",
    "                                  find_max_celltype(gene,adata,partition_key=partition_key),\n",
    "                                  max(one_v_max[gene]), \n",
    "                                  shannon[gene],\n",
    "                                  tau[gene],\n",
    "                                  gini[gene]]\n",
    "    return spe_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ipmc\\.conda\\envs\\scanpypip\\lib\\site-packages\\anndata\\_core\\anndata.py:1094: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    }
   ],
   "source": [
    "spe_matrix = full_spe_matrix(barbry, partition_key='CellType')\n",
    "clf_Deprez_366obs = load(MODEL_PATH + 'clf_train_Deprez_366obs.joblib')\n",
    "spe_type = clf_Deprez_366obs.predict(spe_matrix[['one_v_max','shannon','tau','gini']])\n",
    "spec_dict={'0':'equirep','1':'low','2':'multi-spec','3':'high','4':'unique'}\n",
    "spe_matrix['spe_type'] = pd.Series(spe_type, index = spe_matrix.index).apply(str).replace(spec_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14349, 28)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barbry.varm['ave_celltype_counts_CellType'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>most expressed celltype</th>\n",
       "      <th>one_v_max</th>\n",
       "      <th>shannon</th>\n",
       "      <th>tau</th>\n",
       "      <th>gini</th>\n",
       "      <th>spe_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL669831.5</th>\n",
       "      <td>AL669831.5</td>\n",
       "      <td>AT2</td>\n",
       "      <td>1.56192</td>\n",
       "      <td>0.637306</td>\n",
       "      <td>0.742319</td>\n",
       "      <td>0.391246</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>LINC00115</td>\n",
       "      <td>PNEC</td>\n",
       "      <td>1.22682</td>\n",
       "      <td>0.628045</td>\n",
       "      <td>0.626468</td>\n",
       "      <td>0.399453</td>\n",
       "      <td>equirep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM41C</th>\n",
       "      <td>FAM41C</td>\n",
       "      <td>Deuterosomal</td>\n",
       "      <td>1.12528</td>\n",
       "      <td>0.637185</td>\n",
       "      <td>0.663437</td>\n",
       "      <td>0.393725</td>\n",
       "      <td>equirep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>Precursor</td>\n",
       "      <td>1.46304</td>\n",
       "      <td>0.673486</td>\n",
       "      <td>0.569871</td>\n",
       "      <td>0.219136</td>\n",
       "      <td>equirep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHN1</th>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>Suprabasal N</td>\n",
       "      <td>1.83215</td>\n",
       "      <td>0.519458</td>\n",
       "      <td>0.867355</td>\n",
       "      <td>0.66531</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>Deuterosomal</td>\n",
       "      <td>3.16325</td>\n",
       "      <td>0.538201</td>\n",
       "      <td>0.907686</td>\n",
       "      <td>0.615957</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>Cycling Basal</td>\n",
       "      <td>1.07778</td>\n",
       "      <td>0.414222</td>\n",
       "      <td>0.886558</td>\n",
       "      <td>0.792062</td>\n",
       "      <td>multi-spec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL354822.1</th>\n",
       "      <td>AL354822.1</td>\n",
       "      <td>Ionocyte</td>\n",
       "      <td>1.26802</td>\n",
       "      <td>0.604515</td>\n",
       "      <td>0.729433</td>\n",
       "      <td>0.467284</td>\n",
       "      <td>equirep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004556.1</th>\n",
       "      <td>AC004556.1</td>\n",
       "      <td>Precursor</td>\n",
       "      <td>1.42064</td>\n",
       "      <td>0.578837</td>\n",
       "      <td>0.840537</td>\n",
       "      <td>0.550767</td>\n",
       "      <td>multi-spec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>Deuterosomal</td>\n",
       "      <td>1.93247</td>\n",
       "      <td>0.608708</td>\n",
       "      <td>0.832561</td>\n",
       "      <td>0.461009</td>\n",
       "      <td>multi-spec</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14349 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene most expressed celltype one_v_max   shannon       tau  \\\n",
       "AL669831.5  AL669831.5                     AT2   1.56192  0.637306  0.742319   \n",
       "LINC00115    LINC00115                    PNEC   1.22682  0.628045  0.626468   \n",
       "FAM41C          FAM41C            Deuterosomal   1.12528  0.637185  0.663437   \n",
       "NOC2L            NOC2L               Precursor   1.46304  0.673486  0.569871   \n",
       "PLEKHN1        PLEKHN1            Suprabasal N   1.83215  0.519458  0.867355   \n",
       "...                ...                     ...       ...       ...       ...   \n",
       "AC007325.4  AC007325.4            Deuterosomal   3.16325  0.538201  0.907686   \n",
       "AC007325.2  AC007325.2           Cycling Basal   1.07778  0.414222  0.886558   \n",
       "AL354822.1  AL354822.1                Ionocyte   1.26802  0.604515  0.729433   \n",
       "AC004556.1  AC004556.1               Precursor   1.42064  0.578837  0.840537   \n",
       "AC240274.1  AC240274.1            Deuterosomal   1.93247  0.608708  0.832561   \n",
       "\n",
       "                gini    spe_type  \n",
       "AL669831.5  0.391246         low  \n",
       "LINC00115   0.399453     equirep  \n",
       "FAM41C      0.393725     equirep  \n",
       "NOC2L       0.219136     equirep  \n",
       "PLEKHN1      0.66531         low  \n",
       "...              ...         ...  \n",
       "AC007325.4  0.615957        high  \n",
       "AC007325.2  0.792062  multi-spec  \n",
       "AL354822.1  0.467284     equirep  \n",
       "AC004556.1  0.550767  multi-spec  \n",
       "AC240274.1  0.461009  multi-spec  \n",
       "\n",
       "[14349 rows x 7 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spe_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x21b6ed0e340>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW4AAAFgCAYAAACbqJP/AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAdLUlEQVR4nO3df7RdZX3n8feHoIg/qFADxYRZUJvWAlOxxBR/W1GJ1hqmFY2tJVXaVIpWnTod6MyqrZ3MMEs7VbRQKVVCf0BTfwxpLSiTFlHKMlwVhaBIaiikUIhaFbWlgt/5Yz+R483NzYXck3sf7vu11llnn+/Ze59nn3vzuU+es/dzUlVIkvqx31w3QJL0wBjcktQZg1uSOmNwS1JnDG5J6sz+c92AcVm5cmVdfvnlc90MSdobmar4kO1xf+lLX5rrJkjSWDxkg1uSHqoMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOjPW4E7yxiRbktyQ5OIkj0hySJIrktzc7g8eWf+sJFuT3JTkpJH68Umub8+dk2TKq4kkaSEYW3AnWQL8GrC8qo4FFgGrgTOBTVW1DNjUHpPk6Pb8McBK4Nwki9ruzgPWAsvabeW42i1J8924h0r2Bw5Msj/wSOB2YBWwvj2/Hji5La8CLqmqe6pqG7AVWJHkcOCgqrqmhq/ruWhkG0lacMYW3FX1T8DbgFuBO4CvVdVHgMOq6o62zh3AoW2TJcBtI7vY3mpL2vLkuiQtSOMcKjmYoRd9FPB44FFJXjndJlPUapr6VK+5NslEkokdO3Y80CZLUhfGOa3r84BtVbUDIMkHgKcBdyY5vKruaMMgd7X1twNHjGy/lGFoZXtbnlzfRVWdD5wPsHz5cr8Fubn1Lf9xrpuw1/7Db10/102Q5o1xjnHfCpyQ5JHtLJATgc8BG4E1bZ01wKVteSOwOskBSY5i+BBycxtOuTvJCW0/p45sI0kLzth63FX1iSTvAz4F3At8mqE3/GhgQ5LTGML9lLb+liQbgBvb+mdU1X1td6cDFwIHApe1myQtSGP9BpyqejPw5knlexh631Otvw5YN0V9Ajh21hsoSR3yyklJ6ozBLUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnTG4JakzBrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOmNwS1JnDG5J6ozBLUmdMbglqTMGtyR1ZmzBneRHklw3cvt6kjckOSTJFUlubvcHj2xzVpKtSW5KctJI/fgk17fnzkmScbVbkua7sQV3Vd1UVcdV1XHA8cC3gA8CZwKbqmoZsKk9JsnRwGrgGGAlcG6SRW135wFrgWXttnJc7Zak+W5fDZWcCPxDVf0jsApY3+rrgZPb8irgkqq6p6q2AVuBFUkOBw6qqmuqqoCLRraRpAVnXwX3auDitnxYVd0B0O4PbfUlwG0j22xvtSVteXJ9F0nWJplIMrFjx45ZbL4kzR9jD+4kDwdeAvzlnladolbT1HctVp1fVcuravnixYsfWEMlqRP7osf9QuBTVXVne3xnG/6g3d/V6tuBI0a2Wwrc3upLp6hL0oK0L4L7Fdw/TAKwEVjTltcAl47UVyc5IMlRDB9Cbm7DKXcnOaGdTXLqyDaStODsP86dJ3kk8HzgV0bKZwMbkpwG3AqcAlBVW5JsAG4E7gXOqKr72janAxcCBwKXtZskLUhjDe6q+hbw/ZNqX2Y4y2Sq9dcB66aoTwDHjqONktQbr5yUpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnTG4JakzBrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOmNwS1JnDG5J6ozBLUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjpjcEtSZ8Ya3Ekem+R9ST6f5HNJnprkkCRXJLm53R88sv5ZSbYmuSnJSSP145Nc3547J0nG2W5Jms/G3eN+B3B5VT0ReBLwOeBMYFNVLQM2tcckORpYDRwDrATOTbKo7ec8YC2wrN1WjrndkjRvjS24kxwEPAv4Y4Cq+veq+iqwCljfVlsPnNyWVwGXVNU9VbUN2AqsSHI4cFBVXVNVBVw0so0kLTjj7HH/ILADeG+STye5IMmjgMOq6g6Adn9oW38JcNvI9ttbbUlbnlzfRZK1SSaSTOzYsWN2j0aS5olxBvf+wI8D51XVk4Fv0oZFdmOqceuapr5rser8qlpeVcsXL178QNsrSV0YZ3BvB7ZX1Sfa4/cxBPmdbfiDdn/XyPpHjGy/FLi91ZdOUZekBWlswV1V/wzcluRHWulE4EZgI7Cm1dYAl7bljcDqJAckOYrhQ8jNbTjl7iQntLNJTh3ZRpIWnP3HvP/XAX+W5OHAF4FXMfyx2JDkNOBW4BSAqtqSZANDuN8LnFFV97X9nA5cCBwIXNZukrQgjTW4q+o6YPkUT524m/XXAeumqE8Ax85q4ySpU145KUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnTG4JakzBrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOmNwS1JnDG5J6ozBLUmdMbglqTMGtyR1xuCWpM6MNbiT3JLk+iTXJZlotUOSXJHk5nZ/8Mj6ZyXZmuSmJCeN1I9v+9ma5JwkGWe7JWk+2xc97p+squOqanl7fCawqaqWAZvaY5IcDawGjgFWAucmWdS2OQ9YCyxrt5X7oN2SNC/NxVDJKmB9W14PnDxSv6Sq7qmqbcBWYEWSw4GDquqaqirgopFtJGnBGXdwF/CRJJ9MsrbVDquqOwDa/aGtvgS4bWTb7a22pC1Pru8iydokE0kmduzYMYuHIUnzx/5j3v/Tq+r2JIcCVyT5/DTrTjVuXdPUdy1WnQ+cD7B8+fIp15Gk3o21x11Vt7f7u4APAiuAO9vwB+3+rrb6duCIkc2XAre3+tIp6pK0II0tuJM8Ksljdi4DLwBuADYCa9pqa4BL2/JGYHWSA5IcxfAh5OY2nHJ3khPa2SSnjmwjSQvOOIdKDgM+2M7c2x/486q6PMm1wIYkpwG3AqcAVNWWJBuAG4F7gTOq6r62r9OBC4EDgcvaTZIWpLEFd1V9EXjSFPUvAyfuZpt1wLop6hPAsbPdRknqkVdOSlJnDG5J6ozBLUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnTG4JakzBrckdWZGwZ1k00xqkqTxm3Za1ySPAB4JPC7Jwdz/NWIHAY8fc9skSVPY03zcvwK8gSGkP8n9wf114A/G1yxJ0u5MG9xV9Q7gHUleV1Xv3EdtkiRNY0bfgFNV70zyNODI0W2q6qIxtUuStBszCu4kfwI8AbgO2Pk9kAUY3JK0j830OyeXA0dXVY2zMZKkPZvpedw3AD8wzoZIkmZmpj3uxwE3JtkM3LOzWFUvGUurJEm7NdPg/u1xNkKSNHMzPavko+NuiCRpZmZ6VsndDGeRADwceBjwzao6aFwNkyRNbaY97seMPk5yMrBiHA2SJE3vQc0OWFX/F3ju7DZFkjQTMx0q+ZmRh/sxnNc9o3O6kywCJoB/qqoXJzkE+AuGqzBvAV5WVf/S1j0LOI3hIp9fq6oPt/rxwIXAgcDfAK/3nHJJC9VMe9w/PXI7CbgbWDXDbV8PfG7k8ZnApqpaBmxqj0lyNLAaOAZYCZzbQh/gPGAtsKzdVs7wtSXpIWemY9yvejA7T7IU+ClgHfCfW3kV8Jy2vB64EvivrX5JVd0DbEuyFViR5BbgoKq6pu3zIuBk4LIH0yZJ6t1Mv0hhaZIPJrkryZ1J3t9CeU/eDvwG8J2R2mFVdQdAuz+01ZcAt42st73VlrTlyXVJWpBmOlTyXmAjw7zcS4C/arXdSvJi4K6q+uQMXyNT1Gqa+lSvuTbJRJKJHTt2zPBlJakvMw3uxVX13qq6t90uBBbvYZunAy9pQx2XAM9N8qfAnUkOB2j3d7X1twNHjGy/FLi91ZdOUd9FVZ1fVcuravnixXtqniT1aabB/aUkr0yyqN1eCXx5ug2q6qyqWlpVRzJ86Pi3VfVKhp77mrbaGuDStrwRWJ3kgCRHMXwIubkNp9yd5IQkAU4d2UaSFpyZBvergZcB/wzcAbwUeFAfWAJnA89PcjPw/PaYqtoCbABuBC4HzqiqnXN/nw5cAGwF/gE/mJS0gM10kqnfBdaMnG99CPA2hkDfo6q6kuHsEarqy8CJu1lvHcMZKJPrE8CxM2yrJD2kzbTH/WM7Qxugqr4CPHk8TZIkTWemwb1fkoN3Pmg97pn21iVJs2im4ft7wN8neR/DqXgvY4ohDUnS+M30ysmLkkwwTCwV4Geq6saxtkySNKUZD3e0oDasJWmOPahpXSVJc8fglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOmNwS1JnDG5J6ozBLUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnRlbcCd5RJLNST6TZEuS32n1Q5JckeTmdn/wyDZnJdma5KYkJ43Uj09yfXvunCQZV7slab4bZ4/7HuC5VfUk4DhgZZITgDOBTVW1DNjUHpPkaGA1cAywEjg3yaK2r/OAtcCydls5xnZL0rw2tuCuwTfaw4e1WwGrgPWtvh44uS2vAi6pqnuqahuwFViR5HDgoKq6pqoKuGhkG0lacMY6xp1kUZLrgLuAK6rqE8BhVXUHQLs/tK2+BLhtZPPtrbakLU+uT/V6a5NMJJnYsWPHrB6LJM0XYw3uqrqvqo4DljL0no+dZvWpxq1rmvpUr3d+VS2vquWLFy9+wO2VpB7sk7NKquqrwJUMY9N3tuEP2v1dbbXtwBEjmy0Fbm/1pVPUJWlBGudZJYuTPLYtHwg8D/g8sBFY01ZbA1zaljcCq5MckOQohg8hN7fhlLuTnNDOJjl1ZBtJWnD2H+O+DwfWtzND9gM2VNVfJ7kG2JDkNOBW4BSAqtqSZANwI3AvcEZV3df2dTpwIXAgcFm7SdKCNLbgrqrPAk+eov5l4MTdbLMOWDdFfQKYbnxckhYMr5yUpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnRnnedzz0vH/5aK5bsJe++RbT53rJkiaQ/a4JakzBrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOmNwS1JnFty0rloYnv7Op891E/ba1a+7eq6boHnKHrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqzNiCO8kRSf4uyeeSbEny+lY/JMkVSW5u9wePbHNWkq1Jbkpy0kj9+CTXt+fOSZJxtVuS5rtx9rjvBX69qn4UOAE4I8nRwJnApqpaBmxqj2nPrQaOAVYC5yZZ1PZ1HrAWWNZuK8fYbkma18YW3FV1R1V9qi3fDXwOWAKsAta31dYDJ7flVcAlVXVPVW0DtgIrkhwOHFRV11RVAReNbCNJC84+GeNOciTwZOATwGFVdQcM4Q4c2lZbAtw2stn2VlvSlifXp3qdtUkmkkzs2LFjVo9BkuaLsQd3kkcD7wfeUFVfn27VKWo1TX3XYtX5VbW8qpYvXrz4gTdWkjow1uBO8jCG0P6zqvpAK9/Zhj9o93e1+nbgiJHNlwK3t/rSKeqStCCN86ySAH8MfK6q/s/IUxuBNW15DXDpSH11kgOSHMXwIeTmNpxyd5IT2j5PHdlGkhaccc4O+HTgF4Drk1zXar8JnA1sSHIacCtwCkBVbUmyAbiR4YyUM6rqvrbd6cCFwIHAZe0mSQvS2IK7qj7O1OPTACfuZpt1wLop6hPAsbPXOknql1dOSlJnDG5J6ozBLUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjpjcEtSZwxuSeqMwS1JnTG4JakzBrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktSZ/ee6AZK0N9716381103YK6/9vZ9+wNvY45akzhjcktQZg1uSOjO24E7yniR3JblhpHZIkiuS3NzuDx557qwkW5PclOSkkfrxSa5vz52TJONqsyT1YJw97guBlZNqZwKbqmoZsKk9JsnRwGrgmLbNuUkWtW3OA9YCy9pt8j4laUEZW3BX1VXAVyaVVwHr2/J64OSR+iVVdU9VbQO2AiuSHA4cVFXXVFUBF41sI0kL0r4e4z6squ4AaPeHtvoS4LaR9ba32pK2PLkuSQvWfPlwcqpx65qmPvVOkrVJJpJM7NixY9YaJ0nzyb4O7jvb8Aft/q5W3w4cMbLeUuD2Vl86RX1KVXV+VS2vquWLFy+e1YZL0nyxr4N7I7CmLa8BLh2pr05yQJKjGD6E3NyGU+5OckI7m+TUkW0kaUEa2yXvSS4GngM8Lsl24M3A2cCGJKcBtwKnAFTVliQbgBuBe4Ezquq+tqvTGc5QORC4rN0kacEaW3BX1St289SJu1l/HbBuivoEcOwsNk2SujZfPpyUJM2QwS1JnTG4JakzBrckdcbglqTOGNyS1BmDW5I6Y3BLUmcMbknqjMEtSZ0xuCWpMwa3JHXG4JakzhjcktQZg1uSOmNwS1JnDG5J6ozBLUmdMbglqTMGtyR1xuCWpM4Y3JLUGYNbkjqz/1w3QNLs+Oiznj3XTdhrz77qo3PdhC7Y45akzhjcktQZg1uSOmNwS1JnugnuJCuT3JRka5Iz57o9kjRXugjuJIuAPwBeCBwNvCLJ0XPbKkmaG10EN7AC2FpVX6yqfwcuAVbNcZskaU6kqua6DXuU5KXAyqr6pfb4F4CfqKrXTlpvLbC2PfwR4KZ92tDB44AvzcHrzqWFeMywMI/bY963vlRVKycXe7kAJ1PUdvmLU1XnA+ePvzm7l2SiqpbPZRv2tYV4zLAwj9tjnh96GSrZDhwx8ngpcPsctUWS5lQvwX0tsCzJUUkeDqwGNs5xmyRpTnQxVFJV9yZ5LfBhYBHwnqraMsfN2p05HaqZIwvxmGFhHrfHPA908eGkJOl+vQyVSJIag1uSOmNwPwhJvjHXbZhPkjw+yfvmuh2zZefPN8mRSX5upL48yTlz17J966F+vO3ne8MU9bcked4etv3tJG8aX+um18WHk5rfqup24KWT60n2r6p756BJs+VI4OeAPweoqglgYi4btC8ttOPdqap+a67bsCf2uPdCBm9NckOS65O8vNXPTfKStvzBJO9py6cl+R9z2ebJkrwyyeYk1yV5d5JFSV6V5AtJPprkj5K8q617YbuKdee2oz3TG9ryLyb5yyR/BXwkyaOSvCfJtUk+nWTVyHqXJrm8TR725lk8piOTfD7JBe1n82dJnpfk6iQ3J1kxucfU1jty0q7OBp7Z3ps3JnlOkr/ezWueneTGJJ9N8raR9+sPk3ysvZ8vbvVF7ffm2rb+r4zs5zfa79Jnkpw9W+/JyPtyw8jjN7X34cok/7v9HnwhyTPb89893iTfn+Qj7Wf47iT/mORxu9tnW35C+/l+sr0HT5zN45kli9rv+JZ2fAeO/p4neVH7Xfp4knMm/fyPbu/dF5P82r5stD3uvfMzwHHAkxgui702yVXAVcAzGc41XwIc3tZ/BsM8K/NCkh8FXg48vaq+neRc4JXA7wDHA18D/g749APc9VOBH6uqryT5n8DfVtWrkzwW2Jzk/7X1VgDHAt9ieO8+1Hp5s+GHgFMYpkC4lqHn/AzgJcBvAtfNYB9nAm+qqp2B+5ypVkpyCPCfgCdWVbXj3OlI4NnAE4C/S/JDwKnA16rqKUkOAK5O8hHgicDJDNM5fKvtd1/Zv6pWJHkR8GZg8lDBm4GPV9VbkvwU908tMZ3zgddU1c1JfgI4F3jurLZ67y0DXlFVv5xkA/CzO59I8gjg3cCzqmpbkosnbftE4CeBxwA3JTmvqr69LxptcO+dZwAXV9V9wJ1JPgo8BfgY8IYMMxjeCByc5HCGQNunf5n34ESGgL42CcCBwNOAK6tqB0CSvwB++AHu94qq+kpbfgHwkpHe7SOA/zCy3pfb63yA4f2creDeVlXXt31vATa1UL2eIUyvm6XXAfg68G/ABUk+BIz2yjZU1XeAm5N8keEf+wuAH8v9/3v5PoYAeR7w3qr6FsDIe7gvfKDdf5Lh/ZnsWQwdFarqQ0n+ZbqdJXk0w+/SX7bfLYADZqWls2tbVV3Xlicf+xOBL1bVtvb4Yr73D9aHquoe4J4kdwGHMVzlPXYG996Zag4VquqfkhwMrGTofR8CvAz4RlXdvQ/btycB1lfVWd8tJCcz9B6nci9teC3Dv8aH72a9b056jZ+tqu+Z8Kv1wCZfRDCbFxXcM7L8nZHH32H4vf/usTSPeCA7T/Jhhn+oE1X1S0lWMPwhXA28lvt7llMdY4DXVdWHJ+1z5RTrz6bpjnnn+3Mfu8+Fqdq2u33uB3y1qo574M3cp0Z/T+5j6LzsNOW/72m23Wd56hj33rkKeHkbs1zM0CvZ3J67BnhDW+djwJva/XyyCXhpkkPhu//l/zTwnDam+TCG4YadbmHoocMwre7DZvAaHwZe14KeJE8eee75SQ5JciDDEMHVe3EsD9QtwI+3Nv04cNQU69zN8N/gXVTVSVV1XAvtRwPfV1V/w/AzP25k1VOS7JfkCcAPMsxY+WHg9Pb+kuSHkzwK+Ajw6iSPbPXZHiq5Ezi0/WwPAF78ALa9Cvj51q4XAgdPt8+q+jqwLckpbZskedIsHce+8nngB3P/Zx8vn8O2fA973HvngwzDH59h6I38RlX9c3vuY8ALqmprkn9k6HXPq+CuqhuT/HeGDxH3A74NnAH8NsMfnjuATzFMMwDwR8ClSTYzhP43d9nprn4XeDvw2Rbet3B/YHwc+BOG8eg/n8Xx7Zl4P3BqkusYxsC/MMU6nwXuTfIZ4EJ2P9b/GIb35REMvbQ3jjx3E/BRht75a6rq35JcwPBf8k+192QHcHJVXZ7kOGAiyb8Df8MwHj8r2ucYbwE+AWxjCKaZ+h3g4iSfasdz6wz2+fPAee137GEMn+98Zq8PZB+pqn9N8qvA5Um+xP2dsjnnJe+aVpJfBJZPnvt8vu53PklyIfDXVfWQOcd9pyS3MPz8HtJzcyd5dFV9o/2B/QPg5qr6/blul0MlkrR7v9z+V7aF4UPkd89tcwb2uCWpM/a4JakzBrckdcbglqTOGNzSg5RhLo+nzXU7tPAY3NKD9xyGy7qlfcrg1kNehhkKP5Rhxr0bkrw8yS25f0a8zW3yJ5IsTvL+DDP3XZvk6bvZ55HAa4A3Zpg98JlJto1cDXlQe42HZZhB7u1J/r69/oqRdu0yc6K0Jwa3FoKVwO1V9aSqOha4vNW/XlUrgHcxXN0J8A7g96vqKQwzxV0w1Q6r6hbgD9u6x1XVx4ArgZ9qq6wG3j8yW9yjquppwK8C72m1/8Ywc+JTGGaZe2u79F2alsGtheB64Hmth/3Mqvpaq188cv/Utvw84F3toouNwEFJppyvZAoXAK9qy68C3jvy3MUAVXVV2+djGWYJPLO91pV878yJ0m45V4ke8qrqC0mOB14E/K8Mc1/D9852t3N5P+CpVfWvD+J1rs7wxQLPBhZV1ejXYu1ulsBdZk6U9sQetx7ykjwe+FZV/SnwNtqsgNw/29vLGSbVgmGGvteObHvcNLueavbAixh61++dVN/57UjPYPgSha8x/cyJ0m55ybse8pKcBLyVYS7ubwOnA+9jCNcXMXRgXtFmcnwcw2RCP8rwP9Krquo1u9nvD7f9fIdhfu2PJfkBhlnyDq+qr7b1rmT4w/Bs4CDg1VW1uU1n+3aGM1MC3LLz23ak6RjcWpDGNbtdhm+1WVVVvzBSu5LhK9AW3Bfvajwc45ZmSZJ3Ai9k6MVLY2OPW9qDJK8CXj+pfHVVnTEX7ZEMbknqjGeVSFJnDG5J6ozBLUmdMbglqTMGtyR15v8D7c8oKl8dtc8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.catplot(x='spe_type',kind='count',data=spe_matrix)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get specificity markers - Direct Method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select unique & high to get started"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "spe_genes = spe_matrix.loc[spe_matrix['spe_type'].isin(['unique','high']),:]\n",
    "spe_genes = spe_genes[['gene','most expressed celltype']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>most expressed celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MXRA8</th>\n",
       "      <td>MXRA8</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL691432.2</th>\n",
       "      <td>AL691432.2</td>\n",
       "      <td>PNEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKCZ</th>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DFFB</th>\n",
       "      <td>DFFB</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHG5</th>\n",
       "      <td>PLEKHG5</td>\n",
       "      <td>PNEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL6A1</th>\n",
       "      <td>COL6A1</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL6A2</th>\n",
       "      <td>COL6A2</td>\n",
       "      <td>Fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S100B</th>\n",
       "      <td>S100B</td>\n",
       "      <td>Serous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC011043.1</th>\n",
       "      <td>AC011043.1</td>\n",
       "      <td>PNEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>Deuterosomal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1454 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene most expressed celltype\n",
       "MXRA8            MXRA8              Fibroblast\n",
       "AL691432.2  AL691432.2                    PNEC\n",
       "PRKCZ            PRKCZ                     AT1\n",
       "DFFB              DFFB                     AT1\n",
       "PLEKHG5        PLEKHG5                    PNEC\n",
       "...                ...                     ...\n",
       "COL6A1          COL6A1              Fibroblast\n",
       "COL6A2          COL6A2              Fibroblast\n",
       "S100B            S100B                  Serous\n",
       "AC011043.1  AC011043.1                    PNEC\n",
       "AC007325.4  AC007325.4            Deuterosomal\n",
       "\n",
       "[1454 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spe_genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "spec_gene_groups = spe_genes.groupby('most expressed celltype')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes_spec = {k: len(table.index) for k, table in spec_gene_groups}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_size = pd.DataFrame(barbry.obs.groupby('CellType').size(),columns = ['group_size'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(n_genes_spec.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_plot=group_size.join(n_genes_spec).sort_values('group_size')\n",
    "sns.catplot(x='gene',y='group_size',data=to_plot,s=15,height=11,aspect=1)\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "spec_dict = {k: list(table.index) for k, table in spec_gene_groups}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spec_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.catplot(n_genes_spec,group_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upset Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spec_dict\n",
    "ttest30\n",
    "wilco30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def celltype_to_plot(celltype):\n",
    "    ct_dict = {'spec' : spec_dict[celltype],\n",
    "               'ttest' : ttest30[celltype],\n",
    "               'wilco' : wilco30[celltype]}\n",
    "    ct_dict = upsetplot.from_contents(ct_dict)\n",
    "    return ct_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_plot_AT1 = celltype_to_plot('AT1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = upsetplot.plot(to_plot_AT1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i=0\n",
    "for celltype in spec_dict.keys():\n",
    "    to_plot = celltype_to_plot(celltype)\n",
    "    upsetplot.plot(to_plot)\n",
    "    plt.title(celltype)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = barbry.obs['CellType'].cat.categories\n",
    "\n",
    "for cell in celltypes:\n",
    "    if sum(adata.obs[partition_key] == cell) != 0 :\n",
    "        reduced_adata = adata[adata.obs[partition_key] == cell, :]\n",
    "        mean_expr = np.asarray(reduced_adata.X.mean(axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (scanpypip)",
   "language": "python",
   "name": "scanpypip"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
